src/daft-local-execution/src/intermediate_ops/cross_join.rs

```diff
@@ -0,0 +1,137 @@
+    IntermediateOperatorResult,
+};
+use crate::sinks::cross_join_collect::CrossJoinStateBridgeRef;
+
+struct CrossJoinState {
+    bridge: CrossJoinStateBridgeRef,
+    stream_idx: usize,
+    collect_idx: usize,
+}
+
+impl CrossJoinState {
+    fn new(bridge: CrossJoinStateBridgeRef) -> Self {
+        Self {
+            bridge,
+            stream_idx: 0,
+            collect_idx: 0,
+        }
+    }
+}
+
+impl IntermediateOpState for CrossJoinState {
+    fn as_any_mut(&mut self) -> &mut dyn std::any::Any {
+        self
+    }
+}
+
+pub struct CrossJoinOperator {
+    output_schema: SchemaRef,
+    stream_side: JoinSide,
+    state_bridge: CrossJoinStateBridgeRef,
+}
+
+impl CrossJoinOperator {
+    pub(crate) fn new(
+        output_schema: SchemaRef,
+        stream_side: JoinSide,
+        state_bridge: CrossJoinStateBridgeRef,
+    ) -> Self {
+        Self {
+            output_schema,
+            stream_side,
+            state_bridge,
+        }
+    }
+}
+
+impl IntermediateOperator for CrossJoinOperator {
+    #[instrument(skip_all, name = "CrossJoinOperator::execute")]
+    fn execute(
+        &self,
+        input: Arc<MicroPartition>,
+        mut state: Box<dyn IntermediateOpState>,
+        runtime: &RuntimeRef,
+    ) -> IntermediateOpExecuteResult {
+        let output_schema = self.output_schema.clone();
+
+        if input.is_empty() {
+            let empty = Arc::new(MicroPartition::empty(Some(output_schema)));
+            return Ok((
+                state,
+                IntermediateOperatorResult::NeedMoreInput(Some(empty)),
+            ))
+            .into();
+        }
+
+        let stream_side = self.stream_side;
+
+        runtime
+            .spawn(
+                async move {
+                    let cross_join_state = state
+                        .as_any_mut()
+                        .downcast_mut::<CrossJoinState>()
+                        .expect("CrossJoinState should be used with CrossJoinOperator");
+
+                    let stream_tables = input.get_tables()?;
+                    let collect_tables = cross_join_state.bridge.get_state().await;
+
+                    let stream_tbl = &stream_tables[cross_join_state.stream_idx];
+                    let collect_tbl = &collect_tables[cross_join_state.collect_idx];
+
+                    let (left_tbl, right_tbl) = match stream_side {
+                        JoinSide::Left => (stream_tbl, collect_tbl),
+                        JoinSide::Right => (collect_tbl, stream_tbl),
+                    };
+
+                    let output_tbl = left_tbl.cross_join(right_tbl, stream_side)?;
+
+                    let output_morsel = Arc::new(MicroPartition::new_loaded(
+                        output_schema,
+                        Arc::new(vec![output_tbl]),
+                        None,
+                    ));
+
+                    // increment inner loop index
+                    cross_join_state.collect_idx =
```

Comment:
what if either of these are empty. `collect_tables.len() == 0`

Reply:
Good catch. Empty stream side is already covered by the if statement at the top of execute but I did not cover the collect side case.